### PR TITLE
Extend AsyncClient to support streaming requests / responses

### LINKED ISSUE #317: Extend AsyncClient to support streaming responses

### PR DESCRIPTION
**Pull request by lizan**
- Refactor `AsyncRequestImpl` to `AsyncStreamImpl` and add streaming requests / responses
- Implements `AsyncRequestImpl` based on `AsyncStreamImpl`

Existing `AsyncClientImplTest` covers most of `AsyncStreamImpl`. Working on more tests for streams.

@qiwzhang PTAL.

Closes #317.

### CODE REVIEW

**Comment by mattklein123** [comment]
@lizan I'm OOO tomorrow, will take a look next week when all tests are in place. At a very high level approach is fine.

File: source/common/http/async_client_impl.cc [hunk 1]
```diff
@@ -29,92 +28,92 @@ AsyncClientImpl::~AsyncClientImpl() {
...
-AsyncRequestImpl::AsyncRequestImpl(MessagePtr&& request, AsyncClientImpl& parent,
-                                   AsyncClient::Callbacks& callbacks,
-                                   const Optional<std::chrono::milliseconds>& timeout)
-    : request_(std::move(request)), parent_(parent), callbacks_(callbacks),
-      stream_id_(parent.config_.random_.random()), router_(parent.config_),
-      request_info_(Protocol::Http11), route_(parent_.cluster_.name(), timeout) {
+AsyncClient::Stream* AsyncClientImpl::start(AsyncClient::StreamCallbacks& callbacks,
+                                            const Optional<std::chrono::milliseconds>& timeout) {
+  AsyncStreamImpl* async_request = new AsyncStreamImpl(*this, callbacks, timeout);
+  std::unique_ptr<AsyncStreamImpl> new_request{async_request};
```
Comment (RomanDzhabarov) [inline source/common/http/async_client_impl.cc]: nit: combine into one line
std::unique_ptr<AsyncStreamImpl> new_request(new AsyncStreamImpl(*this, callbacks, timeout));

File: source/common/http/async_client_impl.cc [hunk 2]
```diff
@@ -29,92 +28,92 @@ AsyncClientImpl::~AsyncClientImpl() {
...
 }
 
-void AsyncRequestImpl::cancel() {
-  reset_callback_();
-  cleanup();
+void AsyncStreamImpl::sendHeaders(HeaderMap& headers, bool end_stream) {
+  headers.insertEnvoyInternalRequest().value(Headers::get().EnvoyInternalRequestValues.True);
+  headers.insertForwardedFor().value(parent_.config_.local_info_.address());
+  router_.decodeHeaders(headers, end_stream);
+}
```
Comment (RomanDzhabarov) [inline source/common/http/async_client_impl.cc]: nit: new line in between

**Comment by qiwzhang** [comment]
This is cool.  Just wondering if this stream can be shared by multiple threads.  1) can multiple thread write data to this stream. 2) When response data is received, which thread will be called.

**Comment by mattklein123** [comment]
@qiwzhang no, as written, AsyncClient is thread local and cannot be safely shared between threads. It would be possible to surround it with enough locking to make it safe, but my general recommendation would be to reframe your problem as a thread local problem. I have not yet found a problem that cannot be solved cleanly with this pattern. Happy to discuss this offline with you.

**Comment by qiwzhang** [comment]
@mattklein123 Thanks for your quick response.  In this case, istio/proxy will need to create N gRPC streams to istio/mixer if it has N threads.  I feel it will be better if they can share the same gRPC stream since some transport optimizations of mixer api are done based on the gRPC stream.  We will go with one stream per thread for now to make thing easier, will visit this later if performance is an issue.

**Comment by mattklein123** [comment]
@lizan can you tag me when the tests are finished and you address roman's comments. then I will take a look.

**Comment by lizan** [comment]
@mattklein123 sure, I'm making more changes since I discovered some problems in the tests. In the meanwhile it would be nice if you can take a look on the interfaces (i.e. async_client.h).

**Comment by mattklein123** [comment]
@lizan sure will do.

File: include/envoy/http/async_client.h [hunk 3]
```diff
@@ -50,6 +83,39 @@ class AsyncClient {
...
+    /***
+     * Send trailers. This implicitly ends the stream.
+     * @param trailers supplies the trailers to send
+     */
+    virtual void sendTrailers(HeaderMap& trailers) PURE;
+
+    /***
+     * Close the stream.
+     */
+    virtual void close() PURE;
```
Comment (mattklein123) [inline include/envoy/http/async_client.h]: What does this do exactly? The above methods implicitly send end_stream. Is this a reset? If so can we call it reset() and provide a reset reason?

File: include/envoy/http/async_client.h [hunk 4]
```diff
@@ -62,6 +128,16 @@ class AsyncClient {
...
+
+  /**
+   * Start an HTTP stream asynchronously
+   * @param callbacks the callbacks to be notified of stream status.
+   * @return a stream handle or nullptr if no stream could be started. NOTE: In this case
+   *         onResetStream() has already been called inline. The client owns the stream and
+   *         the handle can be used to send more messages or close the stream.
+   */
+  virtual Stream* start(StreamCallbacks& callbacks,
+                        const Optional<std::chrono::milliseconds>& timeout) PURE;
```
Comment (mattklein123) [inline include/envoy/http/async_client.h]: It's questionable whether timeout makes sense on this interface, since the only reason I could see it really ever being used is for long lived streaming API requests like you are implementing it for. But I don't have a strong opinion and could go either way. I guess since it's optional, we can leave it. (Please update the comments though).

File: include/envoy/http/async_client.h (hunk 3 above)
Comment (lizan) [inline include/envoy/http/async_client.h]: This is comparable to `Request::cancel()` and its implementation are almost resetting stream and cleanup. I can rename it to reset().

The underlying `Filter` (i.e. router_)'s interface doesn't have reset reason, so I didn't provide that here. Do you think it is needed?

File: include/envoy/http/async_client.h (hunk 4 above)
Comment (lizan) [inline include/envoy/http/async_client.h]: I don't have a strong opinion either. I will update comments soon.

**Comment by lizan** [comment]
@mattklein123 Thanks for reviewing it, and sorry I was busy on other things today, will update the PR with tests soon.

File: include/envoy/http/async_client.h (hunk 3 above)
Comment (mattklein123) [inline include/envoy/http/async_client.h]: Sorry you are right, I just looked at the code again. The fact that there is no reason is due to how we are wrapping the router filter to do all of this. I would just rename to reset() for clarity since IMO that makes more sense in the streaming case.

File: include/envoy/http/async_client.h (hunk 4 above)
Comment (lizan) [inline include/envoy/http/async_client.h]: Done

**Comment by lizan** [comment]
@mattklein123 This is ready to review, PTAL. Thanks!

**Comment by mattklein123** [comment]
@RomanDzhabarov please review. I will take a final pass once you are done.

**Comment by RomanDzhabarov** [comment]
I'll take another pass on this today.

File: include/envoy/http/async_client.h [hunk 5]
```diff
@@ -50,18 +83,64 @@ class AsyncClient {
...
 
+  /**
+   * An in-flight HTTP stream.
+   */
+  class Stream {
+  public:
+    virtual ~Stream() {}
+
+    /***
+     * Send headers to the stream.
```
Comment (fengli79) [inline include/envoy/http/async_client.h]: Explicitly says the sendHeaders method cannot be invoked more than once?

File: include/envoy/http/async_client.h [hunk 6]
```diff
@@ -50,18 +83,64 @@ class AsyncClient {
...
+     * @param end_stream supplies whether this is a header only request.
+     */
+    virtual void sendHeaders(HeaderMap& headers, bool end_stream) PURE;
+
+    /***
+     * Send data to the stream.
+     * @param data supplies the data to send.
+     * @param end_stream supplies whether this is the last data.
+     */
+    virtual void sendData(Buffer::Instance& data, bool end_stream) PURE;
```
Comment (fengli79) [inline include/envoy/http/async_client.h]: Explicitly says the sendData can be invoked multiple times if it get streamed.

File: include/envoy/http/async_client.h [hunk 7]
```diff
@@ -50,18 +83,64 @@ class AsyncClient {
...
+
+    /***
+     * Send trailers. This implicitly ends the stream.
+     * @param trailers supplies the trailers to send.
+     */
+    virtual void sendTrailers(HeaderMap& trailers) PURE;
```
Comment (fengli79) [inline include/envoy/http/async_client.h]: Same as sendHeaders, sendTrailer cannot be invoked more than once.

File: include/envoy/http/async_client.h [hunk 8]
```diff
@@ -38,7 +38,40 @@ class AsyncClient {
...
   /**
-   * An in-flight HTTP request
+   * Notifies caller of async HTTP stream status.
+   */
+  class StreamCallbacks {
+  public:
+    virtual ~StreamCallbacks() {}
+
+    /**
+     * Called when the async HTTP stream receives headers.
```
Comment (fengli79) [inline include/envoy/http/async_client.h]: When all headers get received.

File: include/envoy/http/async_client.h [hunk 9]
```diff
@@ -38,7 +38,40 @@ class AsyncClient {
...
+     * @param end_stream whether the response is header only
+     */
+    virtual void onHeaders(HeaderMapPtr&& headers, bool end_stream) PURE;
+
+    /**
+     * Called when the async HTTP stream receives data.
+     * @param data the data received
+     * @param end_stream whether the data is the last data frame
+     */
+    virtual void onData(Buffer::Instance& data, bool end_stream) PURE;
```
Comment (fengli79) [inline include/envoy/http/async_client.h]: It can be invoked multiple times if the data get streamed.

File: include/envoy/http/async_client.h [hunk 10]
```diff
@@ -38,7 +38,40 @@ class AsyncClient {
...
+
+    /**
+     * Called when the async HTTP stream receives trailers.
+     * @param trailers the trailers received.
+     */
+    virtual void onTrailers(HeaderMapPtr&& trailers) PURE;
```
Comment (fengli79) [inline include/envoy/http/async_client.h]: When all trailers get received.

File: include/envoy/http/async_client.h [hunk 11]
```diff
@@ -38,7 +38,40 @@ class AsyncClient {
...
+
+    /**
+     * Called when the async HTTP stream is reset.
+     */
+    virtual void onResetStream() PURE;
```
Comment (fengli79) [inline include/envoy/http/async_client.h]: onReset to matches with the reset() method in Stream interface. Or change it to resetStream().

File: include/envoy/http/async_client.h (hunk 4 above)
Comment (fengli79) [inline include/envoy/http/async_client.h]: Explicitly says if no timeout given, it's infinite.

File: include/envoy/http/async_client.h (hunk 9 above)
Comment (louiscryan) [inline include/envoy/http/async_client.h]: How is stream closure signalled even if no data was delivered? Seems cleaner to separate into a separate callback otherwise closure handling will end up being overloaded into each handle impl

File: include/envoy/http/async_client.h (hunk 6 above)
Comment (louiscryan) [inline include/envoy/http/async_client.h]: See comment above about closing without data. I don't mind collapsing data write with closure here but then this API needs to document how to do that with no data, e.g. using an empty or null buffer?

File: include/envoy/http/async_client.h [hunk 12]
```diff
@@ -50,18 +83,64 @@ class AsyncClient {
...
+    /***
+     * Send trailers. This implicitly ends the stream.
+     * @param trailers supplies the trailers to send.
+     */
+    virtual void sendTrailers(HeaderMap& trailers) PURE;
+
+    /***
+     * Reset the stream.
+     */
+    virtual void reset() PURE;
```
Comment (louiscryan) [inline include/envoy/http/async_client.h]: Does this call trigger StreamCallbacks.onResetStream or does that only occur if the RESET is received from the remote side. If we're thread local then it seems like it should call.

File: include/envoy/http/async_client.h (hunk 6 above)
Comment (louiscryan) [inline include/envoy/http/async_client.h]: ... also what is behavior if a call is made after end_stream has already occurred?

Comment (fengli79) [inline include/envoy/http/async_client.h]: It's programming error. Looks like an error code should be returned.
Without changing the method signature, it should be ignored with an error log?

File: include/envoy/http/async_client.h [hunk 13]
```diff
@@ -38,7 +38,40 @@ class AsyncClient {
   };
 
   /**
-   * An in-flight HTTP request
+   * Notifies caller of async HTTP stream status.
```
Comment (louiscryan) [inline include/envoy/http/async_client.h]: It might be worth having some exposition about the full-duplex nature of streams in the documentation. For instance StreamCallbacks can continue to receive events even if the Stream has had sendXXX(..., end_stream=true) called.

File: source/common/http/async_client_impl.cc [hunk 14]
```diff
@@ -22,120 +21,182 @@ AsyncClientImpl::AsyncClientImpl(const Upstream::ClusterInfo& cluster, Stats::St
...
+  closeLocal(end_stream);
+}
+
+void AsyncStreamImpl::sendTrailers(HeaderMap& trailers) {
+  router_.decodeTrailers(trailers);
+  closeLocal(true);
+}
+
+void AsyncStreamImpl::closeLocal(bool end_stream) {
+  local_closed_ |= end_stream;
```
Comment (louiscryan) [inline source/common/http/async_client_impl.cc]: debug assertion if local already closed and end_stream = true

File: include/envoy/http/async_client.h (hunk 9 above)
Comment (fengli79) [inline include/envoy/http/async_client.h]: To me the StreamDecoderFilter is an interface which already provide all the functionality. The Stream interface looks like an unnecessary wrapper on top of it. Use the StreamDecoderFilter directly by get the router (or a wrapper of router with active count, etc) is more straightforward.

File: source/common/http/async_client_impl.cc [hunk 15]
```diff
@@ -22,120 +21,182 @@ AsyncClientImpl::AsyncClientImpl(const Upstream::ClusterInfo& cluster, Stats::St
...
+  router_.setDecoderFilterCallbacks(*this);
   // TODO: Correctly set protocol in request info when we support access logging.
 }
 
-AsyncRequestImpl::~AsyncRequestImpl() { ASSERT(!reset_callback_); }
+AsyncStreamImpl::~AsyncStreamImpl() { ASSERT(!reset_callback_); }
 
-void AsyncRequestImpl::encodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
-  response_.reset(new ResponseMessageImpl(std::move(headers)));
+void AsyncStreamImpl::encodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
```
Comment (louiscryan) [inline source/common/http/async_client_impl.cc]: @mattklein123  this is some odd nomenclature, why are headers received from the transport in a method called 'encode...' ?

File: include/envoy/http/async_client.h (hunk 4 above)
Comment (louiscryan) [inline include/envoy/http/async_client.h]: What is behavior if timeout is hit, does it implicitly call Stream.reset() and by implication the Callback.onResetStream gets called & on what thread?

File: include/envoy/http/async_client.h (hunk 6 above)
Comment (louiscryan) [inline include/envoy/http/async_client.h]: So we need code to distinguish between programming error (assert) when double local close and/or reset and transport failure local close after remote reset.

File: include/envoy/http/async_client.h (hunk 9 above)
Comment (louiscryan) [inline include/envoy/http/async_client.h]: I think I mostly agree about StreamDecoderFilter but there are many aspects of StreamDecoderFilterCallbacks that people are unlikely to need or want. 

The fact that an explicit 'reset' handler must be installed

https://github.com/lyft/envoy/blob/89de442d82a1727aa7f53604748e953cc27b4849/include/envoy/http/filter.h#L69

does make the API less obvious than this one.

Comment (louiscryan) [inline include/envoy/http/async_client.h]: Hmm... the more I look at this the less convinced I am. The StreamXXXFilter methods are more like the ones I would expect a client to implement and the StreamXXXFilterCallbacks look more like the write interface.

@mattklein123 can probably shed some like on the relationships and why things are implemented this way.

For pure client use-cases I would expect an API more like the one @lizan wrote.

File: include/envoy/http/async_client.h (hunk 4 above)
Comment (mattklein123) [inline include/envoy/http/async_client.h]: This is all single threaded. Yes, when timeout is hit the stream will be reset with callbacks called.

File: include/envoy/http/async_client.h (hunk 6 above)
Comment (mattklein123) [inline include/envoy/http/async_client.h]: A few things:
1) The codec APIs generally follow the http/2 framing format. Since in that API, if you want to close the stream after headers are complete, but without data, you must send an empty data frame along w/ the end stream bit, this is what we do also, for better or worse. Feel free to update the docs to account for this.
2) There are various asserts in different places to account for programmer error, but the main problem with this is due to object lifetime. Since streams may get destroyed right away, you are not guaranteed to actually be able to check vs. just be dealing with memory corruption / crash situation, so in general we just have a contract that says that once certain events happen you can't use the object anymore. I've tried to be very clear about this in codec.h and filter.h, but let me know if it's not clear enough.

File: include/envoy/http/async_client.h (hunk 9 above)
Comment (mattklein123) [inline include/envoy/http/async_client.h]: I think the simple interface that @lizan wrote is the one to use. Here is why this all works this way: The way AsyncClient is implemented is that it basically mimics ActiveStream inside the HTTP connection manager, and instantiates a router filter to do the work. On one hand, this is kind of hack, on the other hand, it's an implementation detail so it doesn't really matter. So, in the end of the data, internally, the async client implements the StreamXXXCallback methods, which is the write interface, and then translates those "writes" into async client request/stream callbacks, which is this interface.

File: include/envoy/http/async_client.h (hunk 12 above)
Comment (mattklein123) [inline include/envoy/http/async_client.h]: It does call, but this is an internal implementation detail of the async client / request, so I don't think we need to expose from this interface.

File: source/common/http/async_client_impl.cc (hunk 15 above)
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: I explained this above, this is an internal implementation detail, the response is being "encoded" into the client, which is then dispatching the appropriate callbacks.

**Comment by mattklein123** [comment]
@lizan there are a huge amount of comments on this review. I'm fine with the overall approach. When everyone's concerns are satisfied or if there are other questions let me know and I can review in detail.

File: include/envoy/http/async_client.h (hunk 5 above)
Comment (lizan) [inline include/envoy/http/async_client.h]: Done

File: include/envoy/http/async_client.h (hunk 6 above)
Comment (lizan) [inline include/envoy/http/async_client.h]: Done, documented that should use sendData with empty buffer to half close the stream.

File: include/envoy/http/async_client.h (hunk 7 above)
Comment (lizan) [inline include/envoy/http/async_client.h]: Done

File: include/envoy/http/async_client.h (hunk 8 above)
Comment (lizan) [inline include/envoy/http/async_client.h]: Done

File: include/envoy/http/async_client.h (hunk 9 above)
Comment (lizan) [inline include/envoy/http/async_client.h]: Done

File: include/envoy/http/async_client.h (hunk 10 above)
Comment (lizan) [inline include/envoy/http/async_client.h]: Done

File: include/envoy/http/async_client.h (hunk 11 above)
Comment (lizan) [inline include/envoy/http/async_client.h]: renamed to onReset

File: source/common/http/async_client_impl.h [hunk 16]
```diff
@@ -152,24 +166,46 @@ class AsyncRequestImpl final : public AsyncClient::Request,
...
-  AsyncClient::Callbacks& callbacks_;
+  AsyncClient::StreamCallbacks& stream_callbacks_;
   const uint64_t stream_id_;
-  std::unique_ptr<MessageImpl> response_;
   Router::ProdFilter router_;
   std::function<void()> reset_callback_;
   AccessLog::RequestInfoImpl request_info_;
   RouteImpl route_;
-  bool complete_{};
+  Buffer::Instance* decoding_buffer_{nullptr};
```
Comment (lizan) [inline source/common/http/async_client_impl.h]: Done

File: test/mocks/http/mocks.h [hunk 17]
```diff
@@ -308,6 +311,24 @@ class MockAsyncClientCallbacks : public AsyncClient::Callbacks {
...
+public:
+  MockAsyncClientStreamCallbacks();
+  ~MockAsyncClientStreamCallbacks();
+
+  void onHeaders(HeaderMapPtr&& headers, bool end_stream) override {
+    onHeaders_(*headers, end_stream);
+  }
+  void onData(Buffer::Instance& data, bool end_stream) override { onData_(data, end_stream); }
+  void onTrailers(HeaderMapPtr&& trailers) override { onTrailers_(*trailers); }
+  void onResetStream() override { onResetStream_(); }
```
Comment (lizan) [inline test/mocks/http/mocks.h]: done

File: test/mocks/http/mocks.h [hunk 18]
```diff
@@ -308,6 +311,24 @@ class MockAsyncClientCallbacks : public AsyncClient::Callbacks {
...
+
+  MOCK_METHOD2(onHeaders_, void(HeaderMap& headers, bool end_stream));
+  MOCK_METHOD2(onData_, void(Buffer::Instance& data, bool end_stream));
```
Comment (lizan) [inline test/mocks/http/mocks.h]: Done

File: test/mocks/http/mocks.h [hunk 19]
```diff
@@ -317,6 +338,19 @@ class MockAsyncClientRequest : public AsyncClient::Request {
 
   MockAsyncClient* client_;
 };
+
+class MockAsyncClientStream : public AsyncClient::Stream {
```
Comment (lizan) [inline test/mocks/http/mocks.h]: This mock is not used, removed. What are others? MockAsyncClientStreamCallbacks is used in the AsyncClientImplTest.

File: source/common/http/async_client_impl.cc (hunk 14 above)
Comment (lizan) [inline source/common/http/async_client_impl.cc]: Done

File: include/envoy/http/async_client.h (hunk 4 above)
Comment (lizan) [inline include/envoy/http/async_client.h]: When timeout, it will call StreamCallbacks.onHeader with :status 504 and onData, the header and data is coming from underlying filters.

See this test:
https://github.com/lyft/envoy/pull/353/files#diff-918cf37446b66e452d1114b88b1e61c4R578

File: include/envoy/http/async_client.h (hunk 12 above)
Comment (lizan) [inline include/envoy/http/async_client.h]: StreamCallbacks.onReset only occur if the RESET is received from the remote.

File: include/envoy/http/async_client.h (hunk 13 above)
Comment (lizan) [inline include/envoy/http/async_client.h]: Done

File: include/envoy/http/async_client.h (hunk 12 above)
Comment (lizan) [inline include/envoy/http/async_client.h]: Now reset() will call StreamCallbacks.onReset

**Comment by lizan** [comment]
@fengli79 @louiscryan @RomanDzhabarov PTAL again, thanks!

File: source/common/http/async_client_impl.cc [hunk 20]
```diff
@@ -22,120 +21,182 @@ AsyncClientImpl::AsyncClientImpl(const Upstream::ClusterInfo& cluster, Stats::St
...
+  while (!active_streams_.empty()) {
+    active_streams_.front()->reset();
   }
 }
 
 AsyncClient::Request* AsyncClientImpl::send(MessagePtr&& request, AsyncClient::Callbacks& callbacks,
                                             const Optional<std::chrono::milliseconds>& timeout) {
-  std::unique_ptr<AsyncRequestImpl> new_request{
-      new AsyncRequestImpl(std::move(request), *this, callbacks, timeout)};
+  AsyncRequestImpl* async_request =
```
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: nit: just initialize this directly in unique_ptr and return front().get() like you do below.

File: source/common/http/async_client_impl.cc [hunk 21]
```diff
@@ -22,120 +21,182 @@ AsyncClientImpl::AsyncClientImpl(const Upstream::ClusterInfo& cluster, Stats::St
...
+void AsyncStreamImpl::sendTrailers(HeaderMap& trailers) {
+  router_.decodeTrailers(trailers);
+  closeLocal(true);
+}
+
+void AsyncStreamImpl::closeLocal(bool end_stream) {
+  ASSERT(!(local_closed_ && end_stream));
+
+  local_closed_ |= end_stream;
+  if (complete())
```
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: nit: braces around all if statements

File: source/common/http/async_client_impl.cc [hunk 22]
```diff
@@ -22,120 +21,182 @@ AsyncClientImpl::AsyncClientImpl(const Upstream::ClusterInfo& cluster, Stats::St
...
+    cleanup();
+}
+
+void AsyncStreamImpl::closeRemote(bool end_stream) {
+  remote_closed_ |= end_stream;
+  if (complete())
```
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: nit: braces around all if statements

File: source/common/http/async_client_impl.h [hunk 23]
```diff
@@ -152,24 +165,47 @@ class AsyncRequestImpl final : public AsyncClient::Request,
...
+  // AsyncClient::Request
+  virtual void cancel() override;
+
+private:
+  void onHeaders(HeaderMapPtr&& headers, bool end_stream) override;
+  void onData(Buffer::Instance& data, bool end_stream) override;
+  void onTrailers(HeaderMapPtr&& trailers) override;
+  void onReset() override;
+  void onComplete();
+  MessagePtr request_;
```
Comment (mattklein123) [inline source/common/http/async_client_impl.h]: nit: newline before this line

File: source/common/http/async_client_impl.h [hunk 24]
```diff
@@ -152,24 +165,47 @@ class AsyncRequestImpl final : public AsyncClient::Request,
...
+
+  // AsyncClient::Request
+  virtual void cancel() override;
+
+private:
+  void onHeaders(HeaderMapPtr&& headers, bool end_stream) override;
+  void onData(Buffer::Instance& data, bool end_stream) override;
+  void onTrailers(HeaderMapPtr&& trailers) override;
+  void onReset() override;
+  void onComplete();
```
Comment (mattklein123) [inline source/common/http/async_client_impl.h]: nit: move to top of section

File: source/common/http/async_client_impl.h [hunk 25]
```diff
@@ -152,24 +165,47 @@ class AsyncRequestImpl final : public AsyncClient::Request,
...
+                               AsyncClient::StreamCallbacks {
+public:
+  AsyncRequestImpl(MessagePtr&& request, AsyncClientImpl& parent, AsyncClient::Callbacks& callbacks,
+                   const Optional<std::chrono::milliseconds>& timeout);
+
+  // AsyncClient::Request
+  virtual void cancel() override;
+
+private:
+  void onHeaders(HeaderMapPtr&& headers, bool end_stream) override;
```
Comment (mattklein123) [inline source/common/http/async_client_impl.h]: prefix overrides with // AsyncClient::StreamCallbacks

File: source/common/http/async_client_impl.cc [hunk 26]
```diff
@@ -22,120 +21,182 @@ AsyncClientImpl::AsyncClientImpl(const Upstream::ClusterInfo& cluster, Stats::St
...
+  closeLocal(end_stream);
 }
 
-void AsyncRequestImpl::onComplete() {
-  complete_ = true;
-  callbacks_.onSuccess(std::move(response_));
-  cleanup();
+void AsyncStreamImpl::sendData(Buffer::Instance& data, bool end_stream) {
+  router_.decodeData(data, end_stream);
+  decoding_buffer_ = &data;
```
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: This is not quite right and is kind of complicated. In short:
1) When you call router_.decodeData(...) it will return a response code which we currently ignore which tells you whether to buffer or not. Buffering is only done in certain cases, for example, if we might retry, or shadow. In this case, the "connection manager" (AsyncClient) is expected to buffer the entire request.
2) In the streaming case, it doesn't make sense to buffer. Therefore, I would probably check the return code of router_.decodeData(...) and if it asks for buffering, throw an exception of some kind to indicate programmer error. 

so basically for the streaming case I think decodingBuffer() should return nullptr, in this normal case, it should return the request body.

File: source/common/http/async_client_impl.cc [hunk 27]
```diff
@@ -22,120 +21,182 @@ AsyncClientImpl::AsyncClientImpl(const Upstream::ClusterInfo& cluster, Stats::St
...
-      new AsyncRequestImpl(std::move(request), *this, callbacks, timeout)};
+  AsyncRequestImpl* async_request =
+      new AsyncRequestImpl(std::move(request), *this, callbacks, timeout);
+  std::unique_ptr<AsyncStreamImpl> new_request{async_request};
 
   // The request may get immediately failed. If so, we will return nullptr.
-  if (!new_request->complete_) {
-    new_request->moveIntoList(std::move(new_request), active_requests_);
-    return active_requests_.front().get();
+  if (!new_request->complete()) {
```
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: If I'm reading the code correctly, I think you actually want to check for remote_close_ here now, not complete_, or, arguably better, you need to set local_close_ in resetStream(). 

The reason for this is that the router might respond right away. If it responds before the request is complete, it will reset. In this case, you should return a null handle.

File: source/common/http/async_client_impl.cc (hunk 26 above)
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: Sorry just to further clarify, buffering *should* continue to work in the AsyncRequestImpl case (non-streaming).

File: source/common/http/async_client_impl.cc (hunk 20 above)
Comment (lizan) [inline source/common/http/async_client_impl.cc]: Having a raw pointer here is to avoid downcasting from `AsyncStreamImpl` to `AsyncRequestImpl` when it returns. `AsyncStreamImpl` is not a subclass of `AsyncClient::Request`.

Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: ok sorry sounds good

File: source/common/http/async_client_impl.cc (hunk 27 above)
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: Sorry just looked at the code in detail (I haven't looked at this stuff in a while). The router will not actually call resetStream() on its own, it relies on the connection manager to reset if the request when the response is complete. (The main server makes the assumption that even in the streaming case, if the response is finished, the request should be reset if it is not finished).

So, in this case, I would probably just check for remote_close_ and if that is set, return a null handle and do not send any more data. You could also replicate the connection manager behavior, and if remote is closed before local is closed, close local and "reset."

Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: P.S., this stuff is pretty complicated. If you want to chat in realtime find me in the public Gitter.

Comment (lizan) [inline source/common/http/async_client_impl.cc]: looked at the code again, check remote_close_ sounds reasonable, done.

File: source/common/http/async_client_impl.cc (hunk 26 above)
Comment (lizan) [inline source/common/http/async_client_impl.cc]: Done

**Comment by lizan** [comment]
@mattklein123 Sorry for the delay, PTAL when you have time, thanks!

File: source/common/http/async_client_impl.cc [hunk 28]
```diff
@@ -105,8 +105,14 @@ void AsyncStreamImpl::sendHeaders(HeaderMap& headers, bool end_stream) {
 }
 
 void AsyncStreamImpl::sendData(Buffer::Instance& data, bool end_stream) {
-  router_.decodeData(data, end_stream);
-  decoding_buffer_ = &data;
+  FilterDataStatus status = router_.decodeData(data, end_stream);
+
+  if (status == FilterDataStatus::StopIterationAndBuffer) {
+    decoding_buffer_ = &data;
```
Comment (mattklein123) [inline source/common/http/async_client_impl.cc]: This is still not right. To probably support buffering you actually need to buffer the data. For example, have Buffer::OwnedImpl, and add each data frame to it. I don't think this is worth supporting, as I don't think there is any reasonable case where for a streaming request we will need buffering. I would refactor the code slightly in such a way where in the base streaming case, you don't support buffering (throw an exception if something results in buffering being needed), but in the non-streaming case you do what we did previously, which is to just return the request's body data.

Comment (lizan) [inline source/common/http/async_client_impl.cc]: Done, refactored it to throw exception in AsyncStreamImpl, and return body in AsyncRequestImpl.

File: source/common/http/async_client_impl.h [hunk 29]
```diff
@@ -164,7 +164,7 @@ class AsyncStreamImpl : public AsyncClient::Stream,
   AccessLog::RequestInfo& requestInfo() override { return request_info_; }
   const std::string& downstreamAddress() override { return EMPTY_STRING; }
   void continueDecoding() override { NOT_IMPLEMENTED; }
-  const Buffer::Instance* decodingBuffer() override { return decoding_buffer_; }
+  const Buffer::Instance* decodingBuffer() override { NOT_IMPLEMENTED; }
```
Comment (mattklein123) [inline source/common/http/async_client_impl.h]: this is going to just crash. Can you throw an actual EnvoyException (which will also end up crashing the process but indirectly) and add a test for this case please.

Comment (lizan) [inline source/common/http/async_client_impl.h]: Done
